File: static/selfplay.js
'use strict';
const $ = (id) => document.getElementById(id);
let es = null;

init();
async function init() {
  const scenarios = await fetch('/api/scenarios').then(r => r.json());
  $('scenario').innerHTML = scenarios
    .map(s => `<option value="${s.key}">${s.name}</option>`).join('');
  $('runBtn').onclick = run;
}

function run() {
  if (es) es.close();
  $('stage').innerHTML = '';
  $('summary').innerHTML = '';
  $('runBtn').disabled = true;
  $('runBtn').textContent = '对弈中…';

  const q = new URLSearchParams({
    scenario: $('scenario').value, level: $('level').value,
    difficulty: $('difficulty').value, turns: $('turns').value,
  });
  es = new EventSource('/api/selfplay?' + q.toString());
  es.onmessage = (e) => handle(JSON.parse(e.data));
  es.onerror = () => finish();
}

function finish() {
  if (es) { es.close(); es = null; }
  $('runBtn').disabled = false;
  $('runBtn').textContent = '重新开始';
}

let lastLearner = null;  // 纠错卡片挂到上一条学习者气泡后

function handle(ev) {
  switch (ev.type) {
    case 'tutor':
      bubble('ai', '🎙️ ' + ev.text); lastLearner = null; break;
    case 'learner':
      lastLearner = bubble('me', '🧑 ' + ev.text); break;
    case 'correction':
      renderCorrection(ev); break;
    case 'summary':
      renderSummary(ev); break;
    case 'judge':
      renderJudge(ev.verdict); break;
    case 'error':
      bubble('ai', '⚠️ ' + ev.msg); finish(); break;
    case 'done':
      finish(); break;
  }
}

function bubble(who, text) {
  const b = document.createElement('div');
  b.className = 'bubble ' + (who === 'me' ? 'me' : 'ai');
  b.textContent = text;
  $('stage').appendChild(b);
  b.scrollIntoView({ behavior: 'smooth', block: 'end' });
  return b;
}

function renderCorrection(ev) {
  const c = document.createElement('div');
  c.className = 'feedback';
  const fixes = ev.items.map(x =>
    `<div class="fix"><span class="old">${x.original}</span> → <span class="new">${x.corrected}</span><div class="lbl">${x.reason}</div></div>`
  ).join('');
  c.innerHTML = `<div class="card"><span class="lbl">表达优化</span>${fixes}` +
    (ev.polished ? `<div class="fix">建议:<span class="new">${ev.polished}</span></div>` : '') +
    `</div>`;
  $('stage').appendChild(c);
  c.scrollIntoView({ behavior: 'smooth', block: 'end' });
}

function renderSummary(ev) {
  const sk = ev.raw_skills || {};
  const dims = [['grammar', '语法'], ['vocabulary', '词汇'],
                ['expression', '表达'], ['communication', '沟通']];
  const bars = dims.map(([k, label]) => {
    const v = sk[k];
    if (v == null) return '';
    return `<div class="bar-row"><span class="k">${label}</span><div class="bar"><i style="width:${v}%"></i></div><span class="v">${v}</span></div>`;
  }).join('');
  const list = (a) => (a && a.length) ? `<ul class="tight">${a.map(x => `<li>${x}</li>`).join('')}</ul>` : '<p class="na">—</p>';
  $('summary').innerHTML = `
    <h2>📊 课后总结</h2>
    <div class="bars">${bars}</div>
    <p class="na">发音/流利无真实人声,故不计入</p>
    <h3>✨ 优秀表达</h3>${list(ev.highlights)}
    <h3>⚠️ 高频错误</h3>${list(ev.frequent_errors)}
    <h3>💡 推荐表达</h3>${list(ev.recommended)}
    <h3>📌 训练建议</h3><div class="advice">${ev.advice || '—'}</div>`;
}

function renderJudge(v) {
  if (!v) return;
  const div = document.createElement('div');
  div.className = 'card';
  div.style.marginTop = '14px';
  div.innerHTML = `<span class="lbl">评委评分</span>
    <div class="score-row">
      <span class="score">自然度 <span class="n">${v.naturalness ?? '?'}</span>/5</span>
      <span class="score">考官在角色 <span class="n">${v.tutor_in_role ?? '?'}</span>/5</span>
    </div>
    <div class="lbl">${v.comment || ''}</div>`;
  $('summary').appendChild(div);
}

File: scripts/selfplay.py
"""自对弈验证(CLI):复用 selfplay.run_selfplay,彩色打印事件流。

用法:python scripts/selfplay.py [scenario] [turns] [difficulty] [learner_level]
    例:python scripts/selfplay.py interview 4 2 A2
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from ai import get_service  # noqa: E402
from selfplay import run_selfplay  # noqa: E402

C = {"tutor": "\033[36m", "learner": "\033[33m", "fix": "\033[31m",
     "ok": "\033[32m", "dim": "\033[90m", "end": "\033[0m", "b": "\033[1m"}


def c(key: str, text: str) -> str:
    return f"{C[key]}{text}{C['end']}"


def main() -> int:
    scenario = sys.argv[1] if len(sys.argv) > 1 else "interview"
    turns = int(sys.argv[2]) if len(sys.argv) > 2 else 4
    difficulty = int(sys.argv[3]) if len(sys.argv) > 3 else 2
    level = sys.argv[4] if len(sys.argv) > 4 else "A2"

    ai = get_service()
    for ev in run_selfplay(ai, scenario, turns=turns, difficulty=difficulty,
                           level=level):
        t = ev["type"]
        if t == "start":
            print(c("b", f"\n=== 自对弈:{ev['scenario']} · 难度L{ev['difficulty']} "
                         f"· 学习者{ev['level']} · {ev['turns']}轮 ==="))
            print(c("dim", "(无真实人声,发音评测跳过)\n"))
        elif t == "tutor":
            print(c("tutor", f"🎙️ 考官: {ev['text']}"))
        elif t == "learner":
            print(c("learner", f"🧑 学习者: {ev['text']}"))
        elif t == "correction":
            for x in ev["items"]:
                print(f"   {c('fix', '✗ ' + x['original'])} → "
                      f"{c('ok', x['corrected'])} {c('dim', '· ' + x['reason'])}")
        elif t == "summary":
            print(c("b", "\n=== 课后总结 ==="))
            print(c("ok", "优秀表达:"), ev.get("highlights") or "—")
            print(c("fix", "高频错误:"), ev.get("frequent_errors") or "—")
            print(c("ok", "推荐表达:"), ev.get("recommended") or "—")
            print("训练建议:", ev.get("advice") or "—")
            sk = ev.get("raw_skills", {})
            print(c("dim", "六维(发音/流利无人声故 None):  "
                           + "  ".join(f"{k}={v}" for k, v in sk.items())))
        elif t == "judge":
            v = ev.get("verdict") or {}
            print(c("b", "\n=== 评委评分 ==="))
            print(f"  自然度 {v.get('naturalness', '?')}/5 · "
                  f"考官在角色 {v.get('tutor_in_role', '?')}/5")
            print(c("dim", "  " + str(v.get("comment", ""))))
    print()
    return 0


if __name__ == "__main__":
    raise SystemExit(main())
